### ISSUE TITLE
Property '$on' is missing in type

### ISSUE THREAD

**Issue by JoshuaM1995** [notiz-dev/nestjs-prisma#65]
I'm using a custom prisma client, and for some reason I'm getting TypeScript errors. They appear to be related to the prisma client type not matching the type in the package.

Here's my custom prisma module:
```
CustomPrismaModule.forRootAsync({
  name: 'PrismaService',
  useFactory: () => prismaClient,
  isGlobal: true,
}),
```

My custom prisma client:
```
import { PrismaClient } from '@prisma/database';
import { userSubscriptionIsActiveModelExtension, userSubscriptionIsActiveResultExtension } from './extensions';

export const prismaClient = new PrismaClient(
  process.env.NODE_ENV !== 'dev'
    ? {
        datasources: {
          db: {
            url: process.env.DATABASE_URL,
          },
        },
      }
    : undefined,
)
  .$extends(userSubscriptionIsActiveResultExtension)
  .$extends(userSubscriptionIsActiveModelExtension);

export type ExtendedPrismaClient = typeof prismaClient;
```

And here's the error:
```
Property '$on' is missing in type 'DynamicClientExtensionThis<TypeMap<Args & { result: { userSubscription: { isActive: () => { needs: { endsAt: true; isLifetime: true; paymentStatus: true; isOnFreeTrial: true; }; compute({ endsAt, isLifetime, paymentStatus, isOnFreeTrial }: { endsAt: Date | null; isLifetime: boolean; paymentStatus: StripePaymentStatu...' but required in type 'PrismaClientLike'.ts(2741)
```

When I comment out the `$extends` in the custom prisma client, the error goes away. I followed the documentation and it was working at one point. I'm not sure if one of the latest updates broke it? Here are my versions:

**@prisma/client: 4.16.1**
**prisma: 4.16.1**
**nestjs-prisma: 0.20.1**

**Comment by marcjulian** [comment]
Hi @JoshuaM1995 have a look at the extensions example which uses `$on` with an extended PrismaClient

Extended Prisma Client

https://github.com/notiz-dev/nestjs-prisma/blob/2b93c9eac9cbab44949240a97948c4f56652f938/examples/extensions/src/prisma.extension.ts#L3-L13

Use `$on`

https://github.com/notiz-dev/nestjs-prisma/blob/2b93c9eac9cbab44949240a97948c4f56652f938/examples/extensions/src/main.ts#L13

**Comment by marcjulian** [comment]
I tested the extensions example with the latest prisma version 4.16.1 and I am seeing the same error as you:

```
src/prisma.extension.ts:25:22 - error TS2339: Property '$on' does not exist on type 'DynamicClientExtensionThis<TypeMap<Args & { result: {}; model: { user: { findByEmail: () => (email: string) => Promise<GetResult<{ id: string; email: string; name: string; }, { [x: string]: () => unknown; }> & {}>; }; }; query: {}; client: {}; }>, TypeMapCb, { ...; }>'.

25 extendedPrismaClient.$on('query', (e) => {});
```

This might be an issue with the latest prisma version. Either `$on` is missing in the types or it is not available any longer for extended PrismaClients.

**Comment by marcjulian** [comment]
@JoshuaM1995 I found the following issue about disallowing `$on` for extended clients: https://github.com/prisma/prisma/issues/19552#issuecomment-1580237625

You can find it also in the release notes from [4.16.0](https://github.com/prisma/prisma/releases/tag/4.16.0)

![CleanShot 2023-06-26 at 10 40 54](https://github.com/notiz-dev/nestjs-prisma/assets/8985933/7ff57705-c297-4de9-ba81-ae20e22ec4cb)

**Comment by JoshuaM1995** [comment]
> @JoshuaM1995 I found the following issue about disallowing `$on` for extended clients: [prisma/prisma#19552 (comment)](https://github.com/prisma/prisma/issues/19552#issuecomment-1580237625)
> 
> You can find it also in the release notes from [4.16.0](https://github.com/prisma/prisma/releases/tag/4.16.0)
> 
> ![CleanShot 2023-06-26 at 10 40 54](https://user-images.githubusercontent.com/8985933/248734897-7ff57705-c297-4de9-ba81-ae20e22ec4cb.png)

Thanks for looking into this. Are there any plans for supporting Prisma 4.16.0 any time soon? I can help contribute if possible.

If not I can try downgrading. That was the first thing I tried, but even after downgrading `@prisma/client` and `prisma` in my package.json files (I'm in a monorepo), I still see it generating the client with 4.16.0. I even tried using npx to generate it with a specific version of the cli, with no luck.

**Comment by JoshuaM1995** [comment]
Also, maybe it would be safer to lock prisma to a patch version in the package.json? @marcjulian 

```
"@prisma/client": "~4.13.0"
```

**Comment by JoshuaM1995** [comment]
Update: I was able to solve this temporarily with PNPM by adding this to my package.json:

```
"pnpm": {
    "overrides": {
      "@prisma/client": "4.15.0",
      "prisma": "4.15.0"
    }
  }
```

**Comment by marcjulian** [comment]
@JoshuaM1995 I have update it to the latest prisma client and added support for extension GA. Please try it out with the latest dev release:

```
npm i nestjs-prisma@0.21.0-dev.3
```

**Comment by marcjulian** [comment]
This has been released with [0.21.0](https://github.com/notiz-dev/nestjs-prisma/releases/tag/v0.21.0)
